tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,284 @@
+        # objective value difference between two iterations is smaller than
+        # tolerance
+        'solve_param_i',  # The parameter index where rotosolve is currently
+        # modifying. Reserved for internal use.
+    ])
+
+
+def _get_initial_state(initial_position, tolerance, expectation_value_function):
+    """Create RotosolveOptimizerResults with initial state of search ."""
+    init_args = {
+        "converged": tf.Variable(False),
+        "num_iterations": tf.Variable(0),
+        "num_objective_evaluations": expectation_value_function.iter,
+        "position": tf.Variable(initial_position),
+        "objective_value": tf.Variable(0.),
+        "last_objective_value": tf.Variable(0.),
+        "tolerance": tolerance,
+        "solve_param_i": tf.Variable(0)
+    }
+    return RotosolveOptimizerResults(**init_args)
+
+
+def minimize(expectation_value_function,
+             initial_position,
+             tolerance=1e-8,
+             max_iterations=50,
+             name=None):
+    """Applies the rotosolve algorithm to minimize a linear combination
+     of quantum measurement expectation values. See arXiv:1903.12166,
+      arXiv:1905.09692
+
+    ### Usage:
+
+    The following example demonstrates the Rotosolve optimizer attempting
+     to find the minimum for two qubit ansatz expectation value.
+
+    ```python
+
+    # In this example we train a circuit perform an XOR operation
+    X = np.asarray([
+        [0, 0],
+        [0, 1],
+        [1, 0],
+        [1, 1],
+    ], dtype=float)
+
+    Y = np.asarray([
+        [-1], [1], [1], [-1]
+    ], dtype=float)
+
+    def convert_to_circuit(input_data):
+        # Encode into quantum datapoint.
+        values = np.ndarray.flatten(input_data)
+        qubits = cirq.GridQubit.rect(1, 2)
+        circuit = cirq.Circuit()
+        for i, value in enumerate(values):
+            if value:
+                circuit.append(cirq.X(qubits[i]))
+        return circuit
+
+    x_circ = tfq.convert_to_tensor([convert_to_circuit(x) for x in X])
+
+    # Create two qubits
+    q0, q1 = cirq.GridQubit.rect(1, 2)
+
+    # Create an anzatz on these qubits.
+    a, b = sympy.symbols('a b') # parameters for the circuit
+    circuit = cirq.Circuit(
+        cirq.rx(a).on(q0),
+        cirq.ry(b).on(q1), cirq.CNOT(control=q0, target=q1))
+
+    # Build the Keras model.
+    model = tf.keras.Sequential([
+        # The input is the data-circuit, encoded as a tf.string
+        tf.keras.layers.Input(shape=(), dtype=tf.string),
+        # The PQC layer returns the expected value of the
+        # readout gate, range [-1,1].
+        tfq.layers.PQC(circuit, cirq.Z(q1)),
+    ])
+
+    def hinge_loss(y_true, y_pred):
+        # Here we use hinge loss as the cost function
+        tf.reduce_mean(tf.cast(1 - y_true * y_pred, tf.float32))
+
+    result = rotosolve_minimizer.minimize(
+        rotosolve_minimizer.function_factory(
+            model, hinge_loss, x_circ, Y), np.random.rand([2])

```

Comment:
I am a little bit afraid of this kind of new and independent optimizer design. I understand that this design is conventional in TF Probability, but Tensorflow Keras provides its own Optimizer class to be inherited (e.g. see AdamOptimizer [here](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/optimizer_v2/adam.py#L32-L281)). 

To me, TFP's optimizer design seems like Keras' optimizer + custom trainer. I think it can't be runnable with keras model fit. I feel it's ok that it works well for saving weights and reloading to use the model after training, but I am not sure for the saving models with keras save function in the future. (it can save optimizer together)

What do you think of, @MichaelBroughton?

Reply:
Making a new Keras Optimizers requires that optimizer to be gradient based. This new approach lets us depart from those standards. Rotosolve doesn't calculate gradients, just like nelder-mead in TFP doesn't use gradients.

You're right that it can't be runnable with keras model fit, but this seems to be the best workaround. I could ask the TFP people about this too. What do you think ?

Reply:
Yes, I've checked gradient-free optimizers around several libraries, it is natural to implement gradient-free optimizers in an independent design. I thought the optimizer is designed enough to support both, but I was wrong. Then I agree with TFP's approach for the seamless ecosystem :) Thank you for your idea!